src/persist/src/file.rs

```diff
@@ -0,0 +1,197 @@
+
+//! In-memory implementations for testing and benchmarking.
+
+use std::collections::{HashMap, HashSet};
+use std::convert::TryInto;
+use std::fs::{File, OpenOptions};
+use std::io::{Read, Seek, SeekFrom, Write as StdWrite};
+use std::path::{Path, PathBuf};
+use std::sync::{Arc, Mutex};
+
+use crate::error::Error;
+use crate::mem::MemSnapshot;
+use crate::persister::{Meta, Persister, Write};
+use crate::{Id, Token};
+
+/// A naive implementation of [Write] and [Meta] backed by a file.
+#[derive(Clone, Debug)]
+pub struct FileStream {
+    dataz: Arc<Mutex<File>>,
+    buf: Vec<u8>,
+}
+
+impl FileStream {
+    /// Create a new FileStream
+    pub fn new<P: AsRef<Path>>(name: P) -> Self {
+        let file = OpenOptions::new()
+            .write(true)
+            .append(true)
+            .create(true)
+            .read(true)
+            .open(name)
+            .expect("failed to create file");
+        FileStream {
+            dataz: Arc::new(Mutex::new(file)),
+            buf: Vec::new(),
+        }
+    }
+}
+
+impl Write for FileStream {
+    fn write_sync(&mut self, updates: &[((String, String), u64, isize)]) -> Result<(), Error> {
+        self.buf.clear();
+        for ((key, val), ts, diff) in updates.iter() {
+            let key_bytes = key.as_bytes();
+            let key_bytes_len = key_bytes.len() as u64;
+            let val_bytes = val.as_bytes();
+            let val_bytes_len = val_bytes.len() as u64;
+
+            self.buf.extend(&key_bytes_len.to_le_bytes());
+            self.buf.extend(key_bytes);
+            self.buf.extend(&val_bytes_len.to_le_bytes());
+            self.buf.extend(val_bytes);
+            self.buf.extend(&ts.to_le_bytes());
+            self.buf.extend(&(*diff as i64).to_le_bytes());
+        }
+
+        self.buf.shrink_to_fit();
+
+        let mut guard = self.dataz.lock()?;
+        guard.write_all(&self.buf)?;
+        // TODO: we will eventually want to change the interface to let us batch up
+        // calls to `sync_all`.
+        guard.sync_all()?;
+        Ok(())
+    }
+}
+
+impl Meta for FileStream {
+    type Snapshot = MemSnapshot;
+
+    fn snapshot(&self) -> Result<Self::Snapshot, Error> {
+        let mut bytes = Vec::new();
+        let mut dataz = Vec::new();
+
+        {
+            let mut guard = self.dataz.lock()?;
+            guard
+                .seek(SeekFrom::Start(0))
+                .expect("failed to seek to beginning");
+            guard.read_to_end(&mut bytes).expect("failed to read file");
+        }
+
+        let mut bytes = &bytes[..];
+        let mut buf = [0u8; 8];
+        while !bytes.is_empty() {
+            // Decode key
+            bytes.read_exact(&mut buf)?;
+            let key_len = u64::from_le_bytes(buf);
+            let mut key: Vec<u8> = vec![0u8; key_len.try_into().unwrap()];
+            bytes.read_exact(&mut key[..])?;
+            let key = String::from_utf8(key).expect("failed to parse key");
```

Comment:
"key was not valid utf-8"?

also return this as an error instead of expect. expect panics which I find tends to be used only when there's nothing the system could possibly do to recover